### ISSUE TITLE
How to correctly annotate a dataset

### ISSUE THREAD

**Issue by Joeyabuki99** [ultralytics/ultralytics#13754]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi guys, I want to use YOLOv8 only for the detection phase, and not for the classification. I have so some custom videos, and I have to annotate them... How many frames should I divide the video into (fps)? Should I than do augmentation? if I have several frames where one object is moving and one is not, should I always annotate the stationary one as well (e.g. 10 seconds at 3 fps, that's 30 frames, and if there is a stationary object of my interest in each frame, do I still have to annotate it each time?) 

I know that in a classification task the same identical bounding box is not good, but for the detection phase?  Thanks

**Comment by knoppmyth** [comment]
Not sure there is a "correct" answer here... This is like asking how many total images should I have or how long should I train. It all depends on what you're attempting to detect and the complexity of the scene. Is what you're trying to detect in a fixed location? Is it in the "wild". In short, you'll need to experiment.

With respect to detection then classification, classification does not use a bounding box. 

Here is a simple example. Lets say you want to detect automobile then classify if the automobile is a sedan or truck. You'd build a general automobile detection with various types of automobiles.

For the classifier, you'd have a directory of just sedans and one with just trucks (no bounding boxes). Once the automobile is detected, you'd extract the image and send it to the classifier to classify it as a sedan or trunk. 

Hope this helps,

km

**Comment by glenn-jocher** [comment]
@knoppmyth hi there,

Great questions! Annotating a dataset for object detection can indeed be a bit nuanced, but let's break it down:

### Frame Selection
When deciding how many frames to annotate from your videos, consider the following:
- **Object Movement**: If your object of interest is moving, you might want to annotate more frames to capture its various positions and appearances. A common practice is to annotate every 5-10 frames, but this can vary based on the object's speed and the video's frame rate.
- **Static Objects**: If the object is stationary, you don't need to annotate every single frame. Instead, you can annotate a few key frames and interpolate the annotations if necessary.

### Annotation Consistency
For detection tasks, it's crucial to annotate the object in every frame where it appears, even if it's stationary. This helps the model learn to detect the object in various contexts and lighting conditions. Consistent annotations across frames ensure the model understands the object's presence and absence accurately.

### Data Augmentation
Data augmentation can be beneficial to increase the diversity of your training data. Techniques like rotation, scaling, flipping, and color adjustments can help the model generalize better. However, be cautious with augmentations that might distort the object's appearance too much.

### Example Workflow
Here's a simplified workflow for annotating and training a YOLOv8 detection model:

1. **Extract Frames**: Extract frames from your video at a chosen interval (e.g., every 5 frames).
2. **Annotate Frames**: Use a tool like [LabelImg](https://github.com/tzutalin/labelImg) to draw bounding boxes around the objects of interest in each frame.
3. **Organize Dataset**: Ensure your dataset follows the YOLO format. You can find more details in the [Dataset Guide](https://docs.ultralytics.com/tasks/detect/#dataset-format).
4. **Train Model**: Use the annotated dataset to train your YOLOv8 model. Here's a quick example in Python:
    ```python
    from ultralytics import YOLO

    # Load a model
    model = YOLO("yolov8n.pt")  # load a pretrained model

    # Train the model
    results = model.train(data="path/to/your/dataset.yaml", epochs=100, imgsz=640)
    ```

### Additional Tips
- **Experimentation**: As you mentioned, there's no one-size-fits-all answer. Experiment with different frame intervals and augmentation techniques to see what works best for your specific use case.
- **Validation**: Always validate your model on a separate set of frames to ensure it generalizes well.

For more detailed guidance, you can refer to the [Ultralytics documentation on object detection](https://docs.ultralytics.com/tasks/detect/).

Hope this helps! Feel free to ask if you have more questions 😊

**Comment by Joeyabuki99** [comment]
> Not sure there is a "correct" answer here... This is like asking how many total images should I have or how long should I train. It all depends on what you're attempting to detect and the complexity of the scene. Is what you're trying to detect in a fixed location? Is it in the "wild". In short, you'll need to experiment.
> 
> With respect to detection then classification, classification does not use a bounding box.
> 
> Here is a simple example. Lets say you want to detect automobile then classify if the automobile is a sedan or truck. You'd build a general automobile detection with various types of automobiles.
> 
> For the classifier, you'd have a directory of just sedans and one with just trucks (no bounding boxes). Once the automobile is detected, you'd extract the image and send it to the classifier to classify it as a sedan or trunk.
> 
> Hope this helps,
> 
> km

Hi, the scenario can be varied, so are the custom videos I recorded. I annotated at 3 fps and did all the annotations correctly...but when I go to test, the detector only sees the moving object...if something is stationary, it doesn't detect it. 
I've done augmentation like: 

Flip: Horizontal
Rotation: Between -15° and +15°
Shear: ±10° Horizontal, ±10° Vertical
Blur: Up to 1.5px
Noise: Up to 0.5% of pixels

Could be this a problem regards the number of raw images of the dataset? SHould I've annotated the videos at more fps? 

P.S. Should I use the `single_cls` parameter as `True` or `False` considering I'm using YOLO only for detection because I want to send the ROI to a separated classifier

**Comment by knoppmyth** [comment]
Do you have images of the object(s) when they are stationary in your dataset?

**Comment by glenn-jocher** [comment]
Hi @Joeyabuki99,

Thank you for the additional information. It sounds like the issue might be related to the balance and diversity of your dataset. Here are a few suggestions to help improve your model's performance:

### Dataset Balance
1. **Increase Stationary Object Samples**: Ensure you have a sufficient number of images with stationary objects. If these are underrepresented, the model might not learn to detect them effectively.
2. **Frame Extraction Rate**: Annotating at a higher frame rate (e.g., 5-10 fps) might help capture more variations of the stationary objects, especially if lighting or background conditions change slightly.

### Data Augmentation
Your current augmentation techniques are good, but you might want to consider adding more variations:
- **Brightness and Contrast Adjustments**: This can help the model learn to detect objects under different lighting conditions.
- **Scale Variations**: Slightly scaling objects can help the model generalize better.

### Training Parameters
Regarding the `single_cls` parameter:
- **Set `single_cls` to `False`**: Since you are using YOLO for detection and plan to classify the ROI separately, you should keep `single_cls` as `False` to allow the model to differentiate between different classes if needed.

### Example Code for Augmentation
Here's an example of how you might include additional augmentations in your training script:
```python
from ultralytics import YOLO

# Load a model
model = YOLO("yolov8n.pt")

# Train the model with additional augmentations
results = model.train(
    data="path/to/your/data.yaml",
    epochs=100,
    imgsz=640,
    augment=True,
    hsv_h=0.015,  # Hue augmentation
    hsv_s=0.7,    # Saturation augmentation
    hsv_v=0.4,    # Value augmentation
    degrees=10.0, # Rotation augmentation
    translate=0.1,# Translation augmentation
    scale=0.5,    # Scale augmentation
    shear=0.1,    # Shear augmentation
    flipud=0.5,   # Vertical flip augmentation
    fliplr=0.5    # Horizontal flip augmentation
)
```

### Monitoring and Validation
- **Validation Set**: Ensure your validation set includes both moving and stationary objects to monitor the model's performance on both scenarios.
- **Tools**: Use tools like TensorBoard or wandb to track your training metrics and visualize how well the model is learning.

If you continue to face issues, please ensure you are using the latest versions of `torch` and `ultralytics`. If the problem persists, providing a minimum reproducible example will help us investigate further. You can find more details on creating a reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Hope this helps! Feel free to reach out if you have more questions 😊
